### PR TITLE
Merge registry into codes

### PR DESCRIPTION
**Pull request by konstin**
## Summary

Currently, we register each rule twice: Once in registry.rs and once in codes.rs. This PR changes this to defining every rule only once in codes.rs. In the process i've refactored and documented `map_codes` a bit.

I've currently only driven this to the point where we have only one macro for registering all the rule, i could push this further by merging codes.rs and registry.rs into a single file (or two, one for registry and one for the helpers) and trimming down the unused `code_to_rule` to just a macro call.

You may find looking at the commits individually helpful commits helpful to see the refactor step by step.

## Test Plan

`cargo test` (and possible CI complaints)

### CODE REVIEW

**Review by MichaReiser — APPROVED** [review]

File: scripts/add_rule.py [hunk 1]
```diff
@@ -122,56 +122,6 @@ def main(*, name: str, prefix: str, code: str, linter: str) -> None:
...
-                fp.write("\n")
-                fp.write(line)
-                fp.write("\n")
-                has_written = True
-
-    assert has_written
-
     text = ""
     with (ROOT_DIR / "crates/ruff/src/codes.rs").open("r") as fp:
         while (line := next(fp)).strip() != f"// {linter}":
```
[the remark this answers is not in the document]
Comment (konstin) [reply scripts/add_rule.py:126]: Thank you!

**Comment by charliermarsh** [comment]
(Please don't merge prior to my review, I'd like to have a chance to read this one.)

File: scripts/add_rule.py [hunk 2]
```diff
@@ -133,17 +132,15 @@ def main(*, name: str, prefix: str, code: str, linter: str) -> None:
...
             " " * 8
-            + f"""({variant}, "{code}") => (RuleGroup::Unspecified, Rule::{name}),\n""",
+            + f"""({variant}, "{code}") => (RuleGroup::Unspecified, {rule}),\n""",
         )
-        lines.sort()
-
         text += "".join(lines)
         text += "\n"
-
+        lines.sort()
```
[the remark this answers is not in the document]
Comment (konstin) [reply scripts/add_rule.py]: oops yes

**Comment by evanrittenhouse** [comment]
Nice work!

File: crates/ruff/src/registry.rs [hunk 3]
```diff
@@ -1,672 +1,16 @@
...
-    rules::flake8_todos::rules::InvalidTodoTag,
-    rules::flake8_todos::rules::MissingTodoAuthor,
-    rules::flake8_todos::rules::MissingTodoLink,
-    rules::flake8_todos::rules::MissingTodoColon,
-    rules::flake8_todos::rules::MissingTodoDescription,
-    rules::flake8_todos::rules::InvalidTodoCapitalization,
-    rules::flake8_todos::rules::MissingSpaceAfterTodoColon,
-);
-
 pub trait AsRule {
```
Comment (charliermarsh) [inline crates/ruff/src/registry.rs:14]: Can / should this and `impl Rule` be in `codes.rs`?

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff_macros/src/map_codes.rs [hunk 4]
```diff
@@ -52,10 +55,12 @@ pub(crate) fn map_codes(func: &ItemFn) -> syn::Result<TokenStream> {
         }
 
         let entry = syn::parse::<Entry>(arm.into_token_stream().into())?;
+        let rule_name = entry.rule.segments.last().unwrap().ident.clone();
         linter_to_rules.entry(entry.linter).or_default().insert(
             entry.code.value(),
             LinterToRuleData {
                 rule_id: entry.rule,
+                rule_name,
                 rule_group_id: entry.group,
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_macros/src/map_codes.rs]: refactored the data into one `Rule` struct

File: crates/ruff/src/registry.rs (hunk 3 above)
Comment (konstin) [reply crates/ruff/src/registry.rs:14]: yes, but i'd do that in a separate PR

**Comment by konstin** [comment]
the code changed a good bit since the last review, would be good if you good give the derive code another look

File: crates/ruff/src/codes.rs [hunk 5]
```diff
@@ -1,6 +1,15 @@
+/// In this module we generate [`Rule`], an enum of all rules, and [`RuleCodePrefix`], an enum of
+/// all rules categories. A rule category is something like pyflakes or flake8-todos. Each rule
+/// category contains all rules and their common prefixes, i.e. everything you can specify in
+/// `--select`. For pylint this is e.g. C0414 and E0118 but also C and E01.
```
Comment (charliermarsh) [inline crates/ruff/src/codes.rs:4]: (Should these be `//! `?)

**Review by charliermarsh — APPROVED** [review]
This looks reasonable but I'd suggest doing some manual testing, just to ensure that the prefixes and selectors work as expected.

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**